ocaml/xapi/xapi_sr.ml

```diff
@@ -778,13 +783,40 @@ let scan ~__context ~sr =
               Db.VDI.get_records_where ~__context
                 ~expr:(Eq (Field "SR", Literal sr'))
             in
+            let string_of_l lst =
+              List.map (fun (_, v) -> v.vDI_uuid) lst |> String.concat ","
+            in
             (* It is sufficient to just compare the refs in two db_vdis, as this
                is what update_vdis uses to determine what to delete *)
             let vdis_ref_equal db_vdi1 db_vdi2 =
-              Listext.List.set_difference (List.map fst db_vdi1)
-                (List.map fst db_vdi2)
-              = []
+              let refs1 = RefSet.of_list (List.map fst db_vdi1) in
+              let refs2 = RefSet.of_list (List.map fst db_vdi2) in
+              (* VDIs that are in db_vdi1 but not in db_vdi2 *)
+              let vdis_removed = RefSet.diff refs1 refs2 in
+              (* VDIs that are in not in db_vdi1 but db_vdi2 *)
+              let vdis_added = RefSet.diff refs2 refs1 in
+
+              ( if not (RefSet.is_empty vdis_removed) then
+                  let removed =
+                    List.filter
+                      (fun (r, _) -> RefSet.mem r vdis_removed)
+                      db_vdi1
+                  in
+                  debug "%s VDIs removed during scan: %s" __FUNCTION__
+                    (string_of_l removed)
+              ) ;
+
+              ( if not (RefSet.is_empty vdis_added) then
+                  let added =
+                    List.filter (fun (r, _) -> RefSet.mem r vdis_added) db_vdi2
+                  in
+                  debug "%s VDIs added during scan: %s" __FUNCTION__
+                    (string_of_l added)
+              ) ;
+
+              RefSet.equal vdis_removed vdis_added
```

Comment:
This is a slightly complicated way of saying that A-B = B-A which is only true when A=B, or `RefSet.equal refs1 refs2`

Reply:
yes agree. As my first version checked if removed and added list were empty I kept the same logic but yes just checking equality of the two sets is logic now. I will modify this.